pip/req/req_set.py

```diff
@@ -228,7 +232,11 @@ def prepare_files(self, finder):
             if not self.ignore_installed and not req_to_install.editable:
                 req_to_install.check_if_exists()
                 if req_to_install.satisfied_by:
-                    if self.upgrade:
+                    # check if we are upgrading and that we don't already have
+                    # an exact version match
+                    if self.upgrade and not (
+                            req_to_install.satisfied_by.as_requirement() ==
+                            req_to_install.req):
```

Comment:
Hmmm, `pip._vendor.pkg_resources.Requirement` overrides `__eq__` and does not override `__ne__`... Suspicious....